### ISSUE TITLE
Split `interfaces/hardhat-ignition.ts` module

### ISSUE THREAD

**Issue by fvictorio** [nomiclabs/hardhat-ignition#55]
There are at least some classes in this module that should live on their own files (mainly `ModuleBuilder` and `Module`). This should make navigating the project a bit easier.

Also, this module is inside the `interfaces` directory, but many of these modules have a lot of logic and are not "implemented" as interfaces anywhere (or at least I can't find that).

**Comment by filippetroviccc** [comment]
I agree this should be more modularized, especially `ModuleBuilder` and `Module`.

The idea behind the folder name `interface` was for user-facing functionality of the library, so probably something like `user-interface` would a bit more self-explanatory but still a bad name.

**Comment by fvictorio** [comment]
I like the way hardhat does it, which is the opposite approach: have an `internal` directory where all the non-public stuff is, and everything outside that directory is considered public API.

In hardhat we want to (eventually) use [subpath exports](https://nodejs.org/api/packages.html#packages_subpath_exports) to do this. Otherwise, you need to have the compilation output at the root of the project*, which is ugly. We didn't use the subpath exports before because they didn't work for node 10, which we support. But Ignition can start with node >=12 support and use that feature.

Anyway, we should start with delimiting the number of exported things and that should give us a better insight into what we need to do here.

\* Because you want to be able to do `import { ... } from "hardhat/builtin-tasks"` instead of `import { ... } from "hardhat/dist/builtin-tasks"`.
